crates/bevy_pbr/src/render/fog.rs

```diff
@@ -0,0 +1,147 @@
+use bevy_app::{App, Plugin};
+use bevy_asset::{load_internal_asset, HandleUntyped};
+use bevy_ecs::{prelude::*, schedule::SystemLabel};
+use bevy_math::{Vec3, Vec4};
+use bevy_reflect::TypeUuid;
+use bevy_render::{
+    extract_component::ExtractComponentPlugin,
+    render_resource::{DynamicUniformBuffer, Shader, ShaderType},
+    renderer::{RenderDevice, RenderQueue},
+    view::ExtractedView,
+    RenderApp, RenderStage,
+};
+
+use crate::{FogFalloff, FogSettings};
+
+/// The GPU-side representation of the fog configuration that's sent as a uniform to the shader
+#[derive(Copy, Clone, ShaderType, Default, Debug)]
+pub struct GpuFog {
+    /// Unsigned int representation of the active fog falloff mode
+    mode: u32,
+    /// Fog color
+    base_color: Vec4,
+    /// The color used for the fog where the view direction aligns with directional lights
+    directional_light_color: Vec4,
+    /// The exponent applied to the directional light alignment calculation
+    directional_light_exponent: f32,
+    /// Allocated differently depending on fog mode.
+    /// See `mesh_view_types.wgsl` for a detailed explanation
+    be: Vec3,
+    /// Allocated differently depending on fog mode.
+    /// See `mesh_view_types.wgsl` for a detailed explanation
+    bi: Vec3,
+}
+
+// Important: These must be kept in sync with `mesh_view_types.wgsl`
+const GPU_FOG_MODE_OFF: u32 = 0;
+const GPU_FOG_MODE_LINEAR: u32 = 1;
+const GPU_FOG_MODE_EXPONENTIAL: u32 = 2;
+const GPU_FOG_MODE_EXPONENTIAL_SQUARED: u32 = 3;
+const GPU_FOG_MODE_ATMOSPHERIC: u32 = 4;
+
+/// Metadata for fog
+#[derive(Default, Resource)]
+pub struct FogMeta {
+    pub gpu_fogs: DynamicUniformBuffer<GpuFog>,
+}
+
+/// Prepares fog metadata and writes the fog-related uniform buffers to the GPU
+pub fn prepare_fog(
+    mut commands: Commands,
+    render_device: Res<RenderDevice>,
+    render_queue: Res<RenderQueue>,
+    mut fog_meta: ResMut<FogMeta>,
+    views: Query<(Entity, Option<&FogSettings>), With<ExtractedView>>,
+) {
+    for (entity, fog) in &views {
+        let gpu_fog = if let Some(fog) = fog {
```

Comment:
Instead of using an if let else and then a match, you can use a match like Some(Linear) => ..., Some(Exponential) => ..., None => GPU_FOG_MODE_OFF.